### ISSUE TITLE
[QIR] Exception when setting UDT named item to an intrinsic callable

### ISSUE THREAD

**Issue by cgranade** [microsoft/qsharp-compiler#955]
**Describe the bug**

When setting a named item of a UDT to an intrinsic callable, the QIR generation rewrite step throws an exception, causing compilation to fail:

**To Reproduce**

```xml
<!-- IntrinsicUdts.csproj -->
<Project Sdk="Microsoft.Quantum.Sdk/0.15.2103133969">

    <PropertyGroup>
      <OutputType>Exe</OutputType>
      <TargetFramework>netcoreapp3.1</TargetFramework>
      <QirGeneration>true</QirGeneration>
      <QscVerbosity>D</QscVerbosity>
    </PropertyGroup>
  
</Project>
```

```qsharp
namespace IntrinsicUdts {
    open Microsoft.Quantum.Intrinsic;
    open Microsoft.Quantum.Canon;

    newtype Options = (
        VerboseMessage: (String -> Unit)
    );

    function DefaultOptions() : Options {
        return Options(
            Ignore<String>
        );
    }

    function IndirectMessage(s : String) : Unit {
        Message(s);
    }

    @EntryPoint()
    operation RunMain() : Unit {
        let options = DefaultOptions()
            w/ VerboseMessage <- IndirectMessage
            // Exception in QIR generation if the above line is replaced by:
            // w/ VerboseMessage <- Message
            ;
    }

}
```

Command to reproduce and observed output:

```
PS> dotnet build
Microsoft (R) Build Engine version 16.9.0-preview-21103-02+198f3f262 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  All projects are up-to-date for restore.
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  (1,1): info: Compiling with command line arguments
      ResponseFiles: obj/qsharp/config/qsc.rsp
      OutputFolder: (null)
      ProjectName: (null)
      EmitDll: False
      QirOutputFolder: (null)
      PerfOutputFolder: (null)
      TrimLevel: 1
      Plugins:
      TargetSpecificDecompositions:
      ExposeReferencesViaTestNames: False
      AdditionalAssemblyProperties:
      RuntimeCapabilityName: (null)
      MakeExecutable: False
      Verbosity: D
      OutputFormat: MsBuild
      Input:
      CodeSnippet: (null)
      WithinFunction: False
      References:
      NoWarn: System.Int32[]
      PackageLoadFallbackFolders:
  (1,1): info: Loaded rewrite steps that are executing as part of the compilation process
      (none)
  (1,1): info: Compiling source files
      C:\Users\cgran\source\scratch\qcvv-scratch\bug-repros\intrinsic-udts\IntrinsicUdts.qs
  (1,1): info: Compiling with referenced assemblies
      C:\Users\cgran\.nuget\packages\microsoft.quantum.targets.interfaces\0.15.2103133969\lib\netstandard2.1\Microsoft.Quantum.Targets.Interfaces.dll
      C:\Users\cgran\.nuget\packages\microsoft.quantum.qsharp.core\0.15.2103133969\lib\netstandard2.1\Microsoft.Quantum.QSharp.Core.dll
      C:\Users\cgran\.nuget\packages\microsoft.quantum.runtime.core\0.15.2103133969\lib\netstandard2.1\Microsoft.Quantum.Runtime.Core.dll
      C:\Users\cgran\.nuget\packages\numsharp\0.20.5\lib\netstandard2.0\NumSharp.Core.dll
      C:\Users\cgran\.nuget\packages\microsoft.quantum.standard\0.15.2103133969\lib\netstandard2.1\Microsoft.Quantum.Standard.dll
      C:\Users\cgran\.nuget\packages\microsoft.quantum.simulators\0.15.2103133969\lib\netstandard2.1\Microsoft.Quantum.Simulation.QCTraceSimulatorRuntime.dll
      C:\Users\cgran\.nuget\packages\microsoft.quantum.simulators\0.15.2103133969\lib\netstandard2.1\Microsoft.Quantum.Simulation.Common.dll
      C:\Users\cgran\.nuget\packages\microsoft.quantum.qsharp.foundation\0.15.2103133969\lib\netstandard2.1\Microsoft.Quantum.QSharp.Foundation.dll
      C:\Users\cgran\.nuget\packages\microsoft.quantum.simulators\0.15.2103133969\lib\netstandard2.1\Microsoft.Quantum.Simulators.dll
  
  ____________________________________________

  Q#: Success! (0 errors, 0 warnings) 

  intrinsic-udts -> C:\Users\cgran\source\scratch\qcvv-scratch\bug-repros\intrinsic-udts\bin\Debug\netcoreapp3.1\intrinsic-udts.dll

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:06.23
```

**System information**

```
dotnet --list-sdks
3.1.201 [C:\Program Files\dotnet\sdk]
3.1.300 [C:\Program Files\dotnet\sdk]
5.0.100 [C:\Program Files\dotnet\sdk]
5.0.200 [C:\Program Files\dotnet\sdk]
6.0.100-preview.1.21103.13 [C:\Program Files\dotnet\sdk]
```

**Comment by bettinaheim** [comment]
@cgranade @swernli With the changes that are currently on main I can no longer reproduce this issue as described - the QIR generation runs fine. What still fails is the linker, since it is expecting an implementation for message body instead of this getting redirected to call into the corresponding runtime library function. I.e. the remaining issue with this one is the same the issue in https://github.com/microsoft/qsharp-compiler/issues/964.

**Comment by bettinaheim** [comment]
Fixed with #990.
